### PR TITLE
Migrate terrain, map concepts into core

### LINKED ISSUE #22: Add terrain visualization as a core concept

### PR DESCRIPTION
**Pull request by TaylorMutch**
This pr reorganizes much of the internal graphics API to be more modular. Taking some inspiration from three.js, the core now internalizes much of the mechanics necessary for 3D plugins to generate terrain. It also reorganizes internal logic to associate a `Plugin` with a `Mesh` (formally thought of as `Renderable`) to allow each `VisualizationPlugin3D` to specify how zonal and other types of statistics may be done.

This resolves #22, #34, #76, and #78.

### CODE REVIEW

**Comment by nikmolnar** [comment]
I'm seeing what I think might be a race condition when trying to create an ENVISION viz:

```python
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/Users/nikmolnar/projects/pyvistas/source/vistas/core/graphics/factory.py", line 30, in run
    self.work()
  File "/Users/nikmolnar/projects/pyvistas/source/vistas/core/graphics/feature/factory.py", line 26, in work
    verts, indices, normals = self.factory.generate_meshes(self.task)
  File "/Users/nikmolnar/projects/pyvistas/source/vistas/core/graphics/feature/factory.py", line 106, in generate_meshes
    if self.use_cache and os.path.exists(self._npz_path):
AttributeError: 'FeatureFactory' object has no attribute 'use_cache'
```

File: source/vistas/core/graphics/feature/factory.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+    def __init__(self, extent, data_src: FeatureDataPlugin, shader=None, plugin=None, initial_zoom=10):
+        if shader is None:
+            shader = FeatureShaderProgram()
+        super().__init__(extent, shader, plugin, initial_zoom)
+
+        self._color_func = None
+        self._render_thread = None
+
+        assert isinstance(data_src, FeatureDataPlugin)
```
Comment (nikmolnar) [inline source/vistas/core/graphics/feature/factory.py]: While technically, this is an [appropriate use](https://wiki.python.org/moin/UsingAssertionsEffectively) of `assert`, I personally prefer _not_ to use it validate parameters for a few reasons: 1) it might get switched off (in production, asserts can be disabled); 2) it's semantically unclear (all asserts are the same, and while a failed assert may provide useful data to a programmer, it's very vague, basically "this is unexpected"); 3) used outside of testing, assert feels temporary, like a first draft.

Instead, I prefer to use `TypeError` when the parameter type is incorrect, or `ValueError` if the parameter is of the correct type, but is wrong (e.g., if your function only takes prime numbers and receives a non-prime number). This is more specific, and allows for an added message to explain further if necessary: `raise TypeError('This function requires a `Foo` because reasons')`.

An exception might be if the assert should not cause a failure, _except_ when in debug; i.e., the code will work just fine in production, but you want to chastise a developer because it *might* be problematic.

\*shrug\* maybe I'm alone here; in which case go ahead and ignore this. Assert just doesn't feel like the right tool to me.

File: source/vistas/core/graphics/vector/geometry.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import numpy
+from OpenGL.GL import *
+
+from vistas.core.graphics.geometry import InstancedGeometry
+
+
+class VectorGeometry(InstancedGeometry):
```
Comment (nikmolnar) [inline source/vistas/core/graphics/vector/geometry.py]: This feels plugin-specific. Is rendering an arrow based on raster inputs generic enough to justify a place in core?

File: plugins/shapefile/main.py [hunk 3]
```diff
@@ -29,21 +28,11 @@ def __init__(self, *args, **kwargs):
 
     def load_data(self):
         self.data_name = self.path.split(os.sep)[-1].split('.')[0]
-
         with fiona.open(self.path, 'r') as shp:
             self.metadata = shp.meta
             projection = Proj(init=self.metadata['crs']['init'])
             self._num_features = len(shp)
-
-        driver = ogr.GetDriverByName('ESRI Shapefile')
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply plugins/shapefile/main.py]: It turns out that `fiona` does with internally with the `bounds` property.

File: plugins/terrain_and_color/main.py [hunk 4]
```diff
@@ -27,6 +22,9 @@
 from vistas.ui.utils import *
 
 
+# Todo - resolve color bug
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply plugins/terrain_and_color/main.py]: New bug that I need to address. The shader program doesn't seem to like it when the `noDataValue` is changed, or that it just isn't being updated appropriately. So just some bug hunting to be done here but since it was a rather minor bug overall I left to do later.

File: source/vistas/core/bounds.py [hunk 5]
```diff
@@ -21,7 +21,7 @@ def center(self) -> Vector3:
         ])
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/bounds.py]: So the `BoundingBox` class could be attributed to other kinds of bounding boxes. An `Extent` could produce a bounding box in 3D space (even though it doesn't now). It also made way for being able to add the `bounding_box.py` for implementing the `BoundingBoxHelper` class, which is a class for rendering the bounding boxes for any `Mesh`.

TL;DR - having both a `bounding_box.py` and a `bounds.py` in the `graphics` module didn't seem to fit very well, and since `BoundingBox` is more of a core concept anyway, not necessarily tied to graphics, I thought it fine to move up.

File: source/vistas/core/graphics/factory.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        super().__init__(*args, **kwargs)
+        self.factory = factory
+        self.task = Task(self.task_name, self.task_description)
+
+    def work(self):
+        """ The work that should be done by this worker thread. """
+
+        raise NotImplementedError
+
+    def run(self):
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/factory.py]: Fair enough, I'll do that.

File: source/vistas/core/graphics/factory.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        self.task.status = Task.RUNNING
+        self.init_event_loop()
+        self.work()
+        self.sync_with_main(post_redisplay)
+        self.task.status = Task.COMPLETE
+
+
+class MeshFactory(Object3D):
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/factory.py]: That is true, perhaps what gets emitted from a `MeshFactory` is a `MeshGroup` which implements the methods from `Object3D` and the `render` method necessary to be rendered. The `MeshFactory` would need to know about the `Scene` that the `MeshGroup` belongs to though, which complicates the interaction with the plugins a bit. Worth testing out.

File: source/vistas/core/graphics/feature/factory.py (hunk 1 above)
Comment (TaylorMutch) [reply source/vistas/core/graphics/feature/factory.py]: I'm good with using `TypeError`, and it would be bad if it were such that a non-`FeatureDataPlugin` was passed to the `FeatureFactory` class. I'll swap it out for a `TypeError`.

File: source/vistas/core/graphics/feature/factory.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        self.data_src = data_src
+
+        self.use_cache = self.data_src is not None
+        if self.use_cache:
+            self._cache = self.data_src.path.replace('.shp', '.ttt')
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/feature/factory.py]: Maybe we don't need the `.shp` extension at all? I'll see about making this more generalized.

File: source/vistas/core/graphics/feature/factory.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            tiles = self.extent.tiles(self.zoom)
+            self._ul = tiles[0]
+            self._br = tiles[-1]
+            self.build()
+
+    def update_features(self, vertices=None, indices=None, normals=None, colors=None):
+        # Update geometry information
+        if vertices is not None and indices is not None and normals is not None:
+            if not self.items:
+                num_indices = num_vertices = len(indices)
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/feature/factory.py]: It may be that there are shared vertices, though I admit I have not taken the time to see about how this could be done.

File: source/vistas/core/graphics/raycaster.py [hunk 10]
```diff
@@ -116,18 +116,32 @@ def __init__(self, origin=None, direction=None, near=None, far=None):
...
         self.ray.direction.normalize()
 
-    def intersect_object(self, obj) -> List[Renderable.Intersection]:
-        intersects = obj.raycast(self)
-        intersects.sort(key=lambda i: i.distance)
+    def intersect_object(self, coords, obj, camera) -> List[Intersection]:
+        """ Retrieve intersections, sorted in ascending distance, to a given Object3D. """
+
+        intersects = []
+        if issubclass(obj.__class__, Object3D):
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/raycaster.py]: I'm not sure what I was thinking here... It probably can be just as you say.

File: source/vistas/core/graphics/vector/geometry.py (hunk 2 above)
Comment (TaylorMutch) [reply source/vistas/core/graphics/vector/geometry.py]: So it's not necessarily based on raster inputs alone. The specification for using a `VectoryGeometry` is flexible enough that as long as there is enough data specified (i.e. 7 floats for each vector), then it doesn't really matter where the data comes from. While it is only used in one visualization currently, I don't think that it is unlikely that it could be used elsewhere. For example, a VectorGeometry could be used to dictate a line of vectors for a wind direction, or be used in something like Larry Mahrt's visualization, where each of the vectors will have their direction altered by some polar degree for each timestep.

File: source/vistas/core/graphics/raycaster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/raycaster.py]: Ah, now I know why. `issubclass` accepts a `class`, not an instance of that class. So, we have to pass it the `__class__` of the instance to make sure it is derived from the `Object3D` class.

[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/raycaster.py]: That's my point. `isinstance` will do the same thing, but you can leave off the `.__class__` part.

```python
class Foo:
   ...

class Bar(Foo):
  ...

>>> b = Bar()
>>> isinstance(b, Foo)
True
```

File: source/vistas/core/bounds.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/bounds.py]: Could you give me an example of using `BoundingBox` in a _non_-graphics context? Or are you planning to combine `Extent` and `BoundingBox` into a single class?

File: source/vistas/core/graphics/overlay.py [hunk 11]
```diff
@@ -72,39 +72,37 @@ def generate_mesh(self):
...
             return self.generate_mesh()
 
-        self.mesh.vertices = numpy.array([
+        self.geometry.vertices = numpy.array([
             [
                 x.position[0], x.position[1], 0,
-                x.position[0]+x.size[0], x.position[1], 0,
-                x.position[0], x.position[1]+x.size[1], 0,
-                x.position[0]+x.size[0], x.position[1]+x.size[1], 0
+                x.position[0] + x.size[0], x.position[1], 0,
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/overlay.py]: Hmm I suppose so. This was following the PEP convention for spacing operators, but if we don't mind bending that rule I'm fine with undoing this.

[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/overlay.py]: The previous formatting is [perfectly allowable in PEP-8](https://www.python.org/dev/peps/pep-0008/#other-recommendations).

> If operators with different priorities are used, consider adding whitespace around the operators with the lowest priority(ies). Use your own judgment; however, never use more than one space, and always have the same amount of whitespace on both sides of a binary operator.

```python
i = i + 1
submitted += 1
x = x*2 - 1
hypot2 = x*x + y*y
c = (a+b) * (a-b)
```

[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/overlay.py]: Granted this case is a little different, since they're separate arguments and not different priority operators. Still, PEP-8 gives you the option of using space around operators, or not, as long as you use the same amount of space on both sides.

File: source/vistas/core/graphics/factory.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/factory.py]: Is a `MeshFactory` even necessary? This class already seems to be a `MeshGroup`, so it's probably fine to rename it and otherwise leave as-is.

File: source/vistas/core/graphics/vector/geometry.py (hunk 2 above)
Comment (nikmolnar) [reply source/vistas/core/graphics/vector/geometry.py]: Ok, sounds good.

File: source/vistas/core/graphics/overlay.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/overlay.py]: Sounds good, I'll undo the formatting change.

File: source/vistas/core/bounds.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/bounds.py]: I had not considered combining `Extent` and `BoundingBox`. They do enough different things that having them separate is good.

One thing I had thought of was having a `BoundingBox` for a selectable area. Here's the use case - for the Terrain & Color plugin, we could define bounding boxes for each of the features, and then allow the features to be highlighted when the user mouses over them. Then all the user has to do is click on that highlighted feature, and then they can view the zonal statistics for that region. This could also be used when defining the bounding box for a user-defined region to perform zonal statistics. All the while, the bounding box does not need to be rendered at all.

So, it's still a graphics-related context... perhaps there isn't a compelling reason to have it external to the graphics module.

File: source/vistas/core/graphics/factory.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/factory.py]: I think that it may be good to have them separate. A `MeshGroup` could be a group of simple meshes, while this `MeshFactory` is what is actually creating the group. A `MeshGroup` could be grass or trees or other 3D features, something that our collaborators have talked about.

That said, I'm fine with calling it a `MeshGroup` for now, and if we need to break it down more granular, then we can do so.

File: source/vistas/core/graphics/feature/factory.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        )
+
+
+class FeatureFactory(MapMeshFactory):
+    """ A MapMeshFactory for handling polygon rendering. """
+
+    worker_class = FeatureFactoryWorker
+
+    def __init__(self, extent, data_src: FeatureDataPlugin, shader=None, plugin=None, initial_zoom=10):
+        if shader is None:
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/feature/factory.py]: Sweet! I didn't know you could do that.

File: source/vistas/core/graphics/feature/factory.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        else:
+            task.progress = 0
+            task.target = self.data_src.get_num_features()
+
+            project = partial(pyproj.transform, self.extent.projection, mercator)   # Our projection method
+
+            tris = []
+            offsets = []
+            offset = 0
+            for i, feature in enumerate(self.data_src.get_features()):
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/feature/factory.py]: Hmm, probably just a debugging leftover. Removed.

File: source/vistas/core/graphics/raycaster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/raycaster.py]: Alrighty, I'll use `isinstance`. I think I mainly was enforcing the fact that it should be a `subclass` of it, because not all `Object3D` objects are renderable.

File: source/vistas/core/graphics/factory.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/factory.py]: I guess I'm seeing how it's a collection of meshes and handles rendering... I don't see where it _creates_ them.

File: source/vistas/core/graphics/feature/factory.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/feature/factory.py]: There will be a slight different in behavior though. In my example, it would choose `FeatureShaderProgram()` any time `shader` evaluates to `False`, not just when it's explicitly `None`. Subtle, but important.

File: source/vistas/core/graphics/raycaster.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/graphics/raycaster.py]: Ah, ok. If your intent is to allow subclasses and _only_ subclasses, then it's perfect as is.

File: source/vistas/core/graphics/factory.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/factory.py]: Mesh geometry (and potentially other long-running pieces) are created in the factory's `MeshWorkerThread`, using the `build` method. The `MeshFactory` starts that factory's `MeshFactoryWorker` thread, which is where the long running geometry creation process happens, which happens in the `work` method. Then the worker communicates back to the factory what should happen with that geometry.

For the `TerrainTileFactory`, it's worker just tosses the geometry back to the factory and the factory adds a `Mesh` to it's `items` attribute, which signals that it can now be rendered into the scene. When the worker is done with all it's assigned work, the `run` method does a refresh on the screen. Having the `run` method use the internal `work` method is a consistent way to make sure that the `event_loop` is started, which prevents problems with the `ElevationSource` throwing errors because no event loop was started in the worker thread.

File: source/vistas/core/graphics/feature/__init__.py [hunk 14]
```diff
@@ -0,0 +1,3 @@
+from .geometry import FeatureGeometry
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/feature/__init__.py]: Actually, I don't know if this is all that helpful here. We don't need to import the whole module, just the classes that will be useful. It could go one way or another. I kind of like it how it is.

File: source/vistas/core/graphics/shader.py [hunk 15]
```diff
@@ -111,40 +115,40 @@ def get_uniform_location(self, name):
     def get_attrib_location(self, name):
         return glGetAttribLocation(self.program, name)
 
-    def uniform1i(self, name, value: int):
```
[the remark this answers is not in the document]
Comment (TaylorMutch) [reply source/vistas/core/graphics/shader.py]: I think this was a debugging thing, I was getting a type error when I thought I shouldn't. I'll add these back in and see if all is well now (it should be).

File: source/vistas/core/bounds.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nikmolnar) [reply source/vistas/core/bounds.py]: Yeah, it seems specific to graphics to me.

**Comment by nikmolnar** [comment]
🍰
